### ISSUE TITLE
Please allow us to define custom file actions

### ISSUE THREAD

**Issue by vesper8** [oneduo/nova-file-manager#83]
On many of my projects I use a custom service which allows me to open local applications, open files in certain applications, launch a browser at a specific url, etc.

For obvious reasons, this only works on my dev.. but that's where I do 100% of my file management anyway so this is perfectly suited to be used with your package.

I'll give you an example of what I'd want to achieve. I'd like to add a custom action so that whenever you highlight a file in your nova-file-manager, there could be a custom new icon added which when clicked would trigger "Open in Finder"

This would call a server function that would receive the file information, including its path, and then that path would be forwarded to a function that would look like this:

```
function myCustomFileAction($file) {
    (new \App\Services\OsxAppService)->openInApp('Finder', $file->path);
}
```

```
<?php

namespace App\Services;

class OsxAppService
{
    public function openInApp($app, $path)
    {
        $command = base_path('bash/openInApp.sh') . " '{$app}' '{$path}' 2>&1";
        exec($command);
    }
}
```

```
#!/bin/sh

app=${1}
path=${2}

if [[ ${app} == "photoshop" ]] ; then
    app="Adobe Photoshop CC 2015"
    cmd="tell application \"${app}\" to open file \"${path}\""
    osascript -e "${cmd}"
elif [[ ${app} == "chrome" ]] ; then
    app="Google Chrome"    
    cmd="open -a \"${app}\" \"${path}\""
    eval $cmd
elif [[ ${app} == "iterm" ]] ; then
    app="iTerm"    
    cmd="open -a \"${app}\" \"${path}\""
    eval $cmd
elif [[ ${app} == "code" ]] ; then
    app="Visual Studio Code"
    cmd="tell application \"${app}\" to open \"${path}\""
    osascript -e "${cmd}"
elif [[ ${app} == "finder" ]] || [[ ${app} == "preview" ]]; then
    app="Finder"
    cmd="tell application \"${app}\" to open (\"${path}\" as POSIX file)"
    osascript -e "${cmd}"
fi

cmd="tell application \"${app}\" to set frontmost to true"
osascript -e "${cmd}"
```

And voila.. super easy.. this works perfectly on Mac.. only on localhost but that's just what I need.

All in all I could see custom file actions being useful in many other scenarios.. but this is the first that comes to mind.

Any chance you could add support for custom file actions pretty please?

Many thanks!! Love your project!

**Comment by Rezrazi** [comment]
Hi @vesper8 

Thank you for your suggestion, however this falls outside the scope of the features we have planned for the package.
This is too specific of a usecase, and we're putting it aside for the moment, we might take a look at it in the future if it shows more interest.

What we may suggest for the time being is that we're planning the addition of Pinia (or Nova) events, these events can be hooked into so as to implement custom logic/actions.

One way we're imagining this for your particular usecase is to emit a `selectedFile` event for instance that will represent a user's selection in the tool, and then allow you to implement your own business logic by adding your own JS/API logic.

This however will not be included in the next 0.7 release, we may suggest taking a look at Pinia state and actions subscribers to implement this feature for the time being.

Thank you !

**Comment by vesper8** [comment]
Thanks for the response @Rezrazi and I like the sound of that idea you've got about implementing onSelect events that I could piggy-back off of.

But that might not work though because that would only allow for one custom action per file and it would occur as soon as you select the file row which would be too early. The ability to define a custom icon with a custom onClick event would be what I'd need. You could add those in the config file, you'd specify an icon class and a click handler and you'd write your own onClick logic.. no need to have the users write additional javascript. On the front-end site in your package you'd just iterate over any custom actions and create a new icon dynamically and emit an event that could be listened to inside the custom click handler.

Maybe what you've got in mind could also work and I'm just not visualizing how it would lead to being able to add one or more custom icons with custom actions.

Anyway loving the new update!  And I've worked with both Pinia and Pintura so these things are definitely great additions in my eyes!

**Comment by vesper8** [comment]
About the new version.. super minor nitpick here.. not even worth opening a new issue over. But I suspect many people will run into this.

I ended up getting the `You are running an outdated version of the package` message, and then I noticed the vendor change and did all of that, removed the vendor folder, removed the config file, removed the old package, added the new one, publish the config.. and I noticed I was still getting the `You are running an outdated version of the package`.. I then looked at the code and noticed the message is being cached for 1 day by default.. so I cleared my cache and it went away. It just seemed odd that I would still see this message for a whole day even after updating everything.
